satellite/metainfo/loop_test.go

```diff
@@ -0,0 +1,334 @@
+		loopCtx, cancel := context.WithCancel(ctx)
+
+		// create 1 normal observer
+		obs1 := newTestObserver(nil)
+
+		var once int64
+		// create another normal observer that will wait before returning during RemoteSegment so we can sync with context cancelation
+		obs2 := newTestObserver(func(ctx context.Context) error {
+			// cancel context during call to obs2.RemoteSegment inside loop
+			fmt.Println("WE ARE CANCELING THE Context")
+			if atomic.AddInt64(&once, 1) == 1 {
+				cancel()
+				<-ctx.Done() // ensure we wait for cancellation to propagate
+			} else {
+				panic("multiple calls to observer after loop cancel")
+			}
+			return nil
+		})
+
+		var group errgroup.Group
+
+		// start loop with cancelable context
+		group.Go(func() error {
+			err := metaLoop.Run(loopCtx)
+			if !errs2.IsCanceled(err) {
+				return errors.New("expected context canceled")
+			}
+			return nil
+		})
+		group.Go(func() error {
+			err := metaLoop.Join(ctx, obs1)
+			if !errs2.IsCanceled(err) {
+				return errors.New("expected context canceled")
+			}
+			return nil
+		})
+		group.Go(func() error {
+			err := metaLoop.Join(ctx, obs2)
+			if !errs2.IsCanceled(err) {
+				return errors.New("expected context canceled")
+			}
+			return nil
+		})
+
+		err := group.Wait()
+		require.NoError(t, err)
+
+		obs3 := newTestObserver(nil)
+		err = metaLoop.Join(ctx, obs3)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "loop closed")
+
+		// expect that obs1 and obs2 each saw fewer than three remote segments
+		fmt.Println(obs1.remoteSegCount)
+		assert.True(t, obs1.remoteSegCount < 3)
+		assert.True(t, obs2.remoteSegCount < 3)
+	})
+}
+
+type testObserver struct {
+	remoteSegCount  int
+	remoteFileCount int
+	inlineSegCount  int
+	uniquePaths     map[string]struct{}
+	onSegment       func(context.Context) error // if set, run this during RemoteSegment()
+}
+
+func newTestObserver(onSegment func(context.Context) error) *testObserver {
+	return &testObserver{
+		remoteSegCount:  0,
+		remoteFileCount: 0,
+		inlineSegCount:  0,
+		uniquePaths:     make(map[string]struct{}),
+		onSegment:       onSegment,
+	}
+}
+
+func (obs *testObserver) RemoteSegment(ctx context.Context, path storj.Path, pointer *pb.Pointer) error {
+	obs.remoteSegCount++
+
+	if _, ok := obs.uniquePaths[path]; ok {
+		// TODO: collect the errors and check in test
+		panic("Expected unique path in observer.RemoteSegment")
+	}
+	obs.uniquePaths[path] = struct{}{}
+
+	if obs.onSegment != nil {
+		return obs.onSegment(ctx)
+	}
+
+	return nil
+}
+
+func (obs *testObserver) RemoteObject(ctx context.Context, path storj.Path, pointer *pb.Pointer) error {
+	obs.remoteFileCount++
+	return nil
+}
+
+func (obs *testObserver) InlineSegment(ctx context.Context, path storj.Path, pointer *pb.Pointer) error {
+	obs.inlineSegCount++
+	if _, ok := obs.uniquePaths[path]; ok {
+		// TODO: collect the errors and check in test
+		panic("Expected unique path in observer.InlineSegment")
```

Comment:
Out of curiosity, the usage of `testing.T` caused any problem? (before it was using `T` now it calls `panic`)

Reply:
Calling `t` concurrently is not thread-safe, it all should happen in the test goroutine.